orttraining/orttraining/test/gradient/gradient_ops_test.cc

```diff
@@ -2099,14 +2125,71 @@ TEST(GradientCheckerTest, SimplifiedLayerNormGrad) {
 TEST(GradientUtilsTest, InPlaceAccumulatorFloat32) {
   OpTester test("InPlaceAccumulator", 1, onnxruntime::kMSDomain);
 
-  test.AddInput<float>("old_sum", {3}, {1, 2, 3});
-  test.AddInput<float>("value", {3}, {4, 5, 6});
+  test.AddInput<float>("old_sum", {3}, {1.f, 2.f, 3.f});
+  test.AddInput<float>("value", {3}, {4.f, 5.f, 6.f});
 
-  test.AddOutput<float>("new_sum", {3}, {5, 7, 9});
+  test.AddOutput<float>("new_sum", {3}, {5.f, 7.f, 9.f});
 
   test.Run();
 }
 
+TEST(GradientUtilsTest, InPlaceAccumulatorV2_CPU) {
+  OpTester test("InPlaceAccumulatorV2", 1, onnxruntime::kMSDomain);
+
+  test.AddInput<float>("old_sum", {3}, {1.f, 2.f, 3.f});
+  test.AddInput<float>("value", {3}, {4.f, 5.f, 6.f});
+  test.AddOutput<bool>("updated", {1}, {true});
+  test.AddOutput<float>("new_sum", {3}, {5.f, 7.f, 9.f});
+
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kCudaExecutionProvider});
+}
+
+TEST(GradientUtilsTest, InPlaceAccumulatorV2Overwrite) {
+  OpTester test("InPlaceAccumulatorV2", 1, onnxruntime::kMSDomain);
+
+  test.AddInput<float>("old_sum", {3}, {1.f, 2.f, 3.f});
+  test.AddInput<float>("value", {3}, {4.f, 5.f, 6.f});
+  test.AddInput<bool>("overwrite", {1}, {true});
+  test.AddOutput<bool>("updated", {1}, {true});
+  test.AddOutput<float>("new_sum", {3}, {4.f, 5.f, 6.f});
+
+  test.Run();
+}
+
+#if defined(USE_CUDA)
+// TODO: Add rocm kernel defs
+TEST(GradientUtilsTest, InPlaceAccumulatorV2_GPU) {
```

Comment:
why is this separate from InPlaceAccumulatorV2_CPU? could one of them just run with all enabled EPs?

Reply:
@edgchen1 That's because this op strictly overwrites the input buffer. Because of that, the same op tester can't be used to run the test consecutively for two EPs since the assertion won't hold for the second run.

Reply:
I see, thanks for the explanation. and InPlaceAccumulatorV2Overwrite works with CPU and CUDA EPs because the input `old_sum` does not affect the output?
maybe a comment would be helpful.
also if we only want one EP to run, consider explicitly specifying the EP that should be run (OpTester::Run has another parameter for that), instead of which EP types to exclude as more EPs could support it later.

Reply:
The `InPlaceAccumulatorV2Overwrite` test will overwrite the accumulation buffer (input 0) without performing accumulation. So, if two tests were to run back to back (example):
```
accumulation_buffer = 1, new_value = 2 => accumulation_buffer = 2
accumulation_buffer = 2, new_value = 2 => accumulation_buffer = 2
```

where as if the overwrite flag is not provided, the accumuation occurs. So, if two tests were to run back to back:
```
accumulation_buffer = 1, new_value = 2 => accumulation_buffer = 3
accumulation_buffer = 3, new_value = 2 => accumulation_buffer = 5
```

I'll make the changes to pass in the EP explicitly.